ktor-network/common/test/io/ktor/network/sockets/tests/UDPSocketTest.kt

```diff
@@ -140,89 +128,124 @@ class UDPSocketTest : CoroutineScope {
 
-        assertEquals(1, done)
+        assertEquals(1, done.value)
 
         socket.socketContext.join()
         assertTrue(socket.isClosed)
     }
 
     @Test
-    fun testOutgoingInvokeOnClosed() = runBlocking {
+    fun testOutgoingInvokeOnClosed() = testUdpSockets { selector ->
         val socket: BoundDatagramSocket = aSocket(selector)
             .udp()
             .bind()
 
         socket.outgoing.close(AssertionError())
 
-        var done = 0
         socket.outgoing.invokeOnClose {
             done += 1
             assertTrue(it is AssertionError)
         }
 
-        assertEquals(1, done)
+        assertEquals(1, done.value)
 
         socket.socketContext.join()
         assertTrue(socket.isClosed)
     }
 
     @Test
     fun testBind() {
-        val socketBuilder: UDPSocketBuilder = aSocket(ActorSelectorManager(Dispatchers.IO)).udp()
+        val socketBuilder: UDPSocketBuilder = aSocket(SelectorManager()).udp()
         val socket = socketBuilder.bind()
         val port = socket.localAddress.port
         socket.close()
 
         repeat(1024) {
             try {
                 socketBuilder
-                    .bind(InetSocketAddress("0.0.0.0", port))
+                    .bind(NetworkAddress("0.0.0.0", port))
                     .close()
-            } catch (_: BindException) {
+            } catch (_: Exception /* TODO: BindException */) {
                 // Don't confuse with: Socket Exception: Already bound
             }
         }
     }
+
+    @Test
+    fun testSendReceive(): Unit = testUdpSockets { selector ->
+        aSocket(selector)
+            .udp()
+            .bind(NetworkAddress("127.0.0.1", 8000)) {
+                reuseAddress = true
+                reusePort = true
+            }
+            .use { socket ->
+                // Send messages to localhost
+                launch {
+                    val address = NetworkAddress("127.0.0.1", 8000)
+                    repeat(10) {
+                        val bytePacket = buildPacket { append("hello") }
+                        val data = Datagram(bytePacket, address)
+                        socket.send(data)
+                    }
+                }
+
+                // Receive messages from localhost
+                repeat(10) {
+                    val incoming = socket.receive()
+                    assertEquals("hello", incoming.packet.readText())
+                }
+            }
+    }
+}
+
+private fun testUdpSockets(block: suspend CoroutineScope.(SelectorManager) -> Unit) {
+    if (!PlatformUtils.IS_JVM && !PlatformUtils.IS_NATIVE) return
+    testSuspend {
+        withTimeout(1000) {
+            SelectorManager().use { selector ->
+                block(selector)
+            }
+        }
+    }
 }
 
 private val OS_NAME: String
     get() {
-        val os = System.getProperty("os.name", "unknown").toLowerCase()
+        val os = "unknown" // TODO: System.getProperty("os.name", "unknown").toLowerCase()
```

Comment:
For now I just put unknown here until I know whether this is needed for native as well. See my comment on the `testBroadcastFails` test where OS_NAME is used.

Reply:
Fixed: https://github.com/ktorio/ktor/pull/2129/commits/09ca26aad261e13af438dc26110f71225046d7d9